mantidproject/mantid
# Enable FilterEvents work with splitters out of time range

Description of work.

**To test:**

Run the following script, which split a workspace into several ones by time first. And then a splitter workspace is generated from the original workspace, such that for each child workspace, some of the splitters are not in its time range.  The goal of this PR is to make sure that FilterEvents creates the empty child workspace if the corresponding splitter is out of time range of the workspace that it operates on.

```
# split original workspace into continuous workspaces
Load(Filename='/SNS/VULCAN/IPTS-13183/data/VULCAN_68607_event.nxs', OutputWorkspace='ws')
GenerateEventsFilter(InputWorkspace='ws', OutputWorkspace='TimeSlicer', InformationWorkspace='linfo', StartTime='0', TimeInterval='3000', LogValueInterval=500)
FilterEvents(InputWorkspace='ws', SplitterWorkspace='TimeSlicer', OutputWorkspaceBaseName='workspace', InformationWorkspace='linfo', GroupWorkspaces=True,
    OutputWorkspaceIndexedFrom1=True) 
    
ws = mtd['ws']
print 'original workspace: run start = {0}, run end = {1}'.format(ws.run().getProperty('proton_charge').times[0], ws.run().getProperty('proton_charge').times[-1])
ws4 = mtd['workspace_4']
print 'child 1 workspace : run start = {1}, run end = {1}'.format(ws4.run().getProperty('proton_charge').times[0], ws4.run().getProperty('proton_charge').times[-1])
# replace run start
r4 = ws4.run()
run_start = ws4.run().getProperty('proton_charge').times[0]
r4.addProperty('run_start', str(run_start), True)

    
# generate a splitters workspace from log value
is_matrix = False
if False:
    is_matrix = True
    GenerateEventsFilter(InputWorkspace='ws', OutputWorkspace='LoadFrameSlicer', InformationWorkspace='linfo', StartTime='0', TimeInterval='700', LogName='loadframe.furnace1', LogValueInterval=100,
        FastLog=True)
    #FilterEvents(InputWorkspace='workspace_{0}'.format(4), SplitterWorkspace='LoadFrameSlicer', OutputWorkspaceBaseName='matrix_output{0}'.format(4), InformationWorkspace='linfo', 
        #GroupWorkspaces=True, OutputWorkspaceIndexedFrom1=False)
else:
    GenerateEventsFilter(InputWorkspace='ws', OutputWorkspace='LoadFrameSlicer', InformationWorkspace='linfo', StartTime='0', TimeInterval='700', LogName='loadframe.furnace1', LogValueInterval=100,
        FastLog=False)
    #FilterEvents(InputWorkspace='workspace_{0}'.format(4), SplitterWorkspace='LoadFrameSlicer', OutputWorkspaceBaseName='splitter_output{0}'.format(4), InformationWorkspace='linfo', 
     #   GroupWorkspaces=True, OutputWorkspaceIndexedFrom1=False) 


# Filter
for index in  [1, 2, 3, 4, 5, 6]:
    if is_matrix:
        output_base = 'matrix{0}'.format(index)
    else:
        output_base = 'splitter{0}'.format(index)
    FilterEvents(InputWorkspace='workspace_{0}'.format(index), SplitterWorkspace='LoadFrameSlicer', OutputWorkspaceBaseName='output{0}'.format(index), InformationWorkspace='linfo', 
        GroupWorkspaces=True, OutputWorkspaceIndexedFrom1=True) 
```



Fixes #19140.




---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
